icu4c/source/common/ucurr.cpp

```diff
@@ -1009,113 +993,80 @@ collectCurrencyNames(const char* locale,
-                if (uhash_get(currencyIsoCodes, iso) != nullptr) {
-                    ures_close(names);
-                    continue;
-                } else {
-                    uhash_put(currencyIsoCodes, iso, iso, &ec3); 
-                }
+            iso = (char*)ures_getKey(names.getAlias());
+            if (localeLevel != 0 && uhash_get(currencyIsoCodes.getAlias(), iso) != nullptr) {
+                continue;
             }
+            uhash_put(currencyIsoCodes.getAlias(), iso, iso, &ec3); 
             // Add currency symbol.
-            (*currencySymbols)[*total_currency_symbol_count].IsoCode = iso;
-            (*currencySymbols)[*total_currency_symbol_count].currencyName = (char16_t*)s;
-            (*currencySymbols)[*total_currency_symbol_count].flag = 0;
-            (*currencySymbols)[(*total_currency_symbol_count)++].currencyNameLen = len;
+            (*currencySymbols)[(*total_currency_symbol_count)++] = {iso, (char16_t*)s, len, 0};
+
             // Add equivalent symbols
             if (currencySymbolsEquiv != nullptr) {
                 UnicodeString str(true, s, len);
                 icu::EquivIterator iter(*currencySymbolsEquiv, str);
                 const UnicodeString *symbol;
                 while ((symbol = iter.next()) != nullptr) {
-                    (*currencySymbols)[*total_currency_symbol_count].IsoCode = iso;
-                    (*currencySymbols)[*total_currency_symbol_count].currencyName =
-                        const_cast<char16_t*>(symbol->getBuffer());
-                    (*currencySymbols)[*total_currency_symbol_count].flag = 0;
-                    (*currencySymbols)[(*total_currency_symbol_count)++].currencyNameLen = symbol->length();
+                    (*currencySymbols)[(*total_currency_symbol_count)++]
+                        = {iso, const_cast<char16_t*>(symbol->getBuffer()), symbol->length(), 0};
                 }
             }
 
             // Add currency long name.
-            s = ures_getStringByIndex(names, UCURR_LONG_NAME, &len, &ec2);
-            (*currencyNames)[*total_currency_name_count].IsoCode = iso;
+            s = ures_getStringByIndex(names.getAlias(), UCURR_LONG_NAME, &len, &ec2);
             char16_t* upperName = toUpperCase(s, len, locale);
-            (*currencyNames)[*total_currency_name_count].currencyName = upperName;
-            (*currencyNames)[*total_currency_name_count].flag = NEED_TO_BE_DELETED;
-            (*currencyNames)[(*total_currency_name_count)++].currencyNameLen = len;
+
+            (*currencyNames)[(*total_currency_name_count)++] = {iso, upperName, len, NEED_TO_BE_DELETED};
 
             // put (iso, 3, and iso) in to array
             // Add currency ISO code.
-            (*currencySymbols)[*total_currency_symbol_count].IsoCode = iso;
-            (*currencySymbols)[*total_currency_symbol_count].currencyName = (char16_t*)uprv_malloc(sizeof(char16_t)*3);
+            upperName = (char16_t*)uprv_malloc(sizeof(char16_t)*3);
```

Comment:
There's nothing to gain by reusing the `upperName` variable, instead allocating a new one, like eg. `char16_t* isoUChars` will make it clearer to the reader what's going on.